R/cfr_time_varying.R

```diff
@@ -0,0 +1,229 @@
+#' a probability mass function parameterised by time; i.e. \eqn{f(t)} which
+#' gives the probability a case has a known outcomes (usually, death) at time
+#' \eqn{t}, parameterised with disease-specific parameters before it is supplied
+#' here.
+#'
+#' @references
+#' Nishiura, H., Klinkenberg, D., Roberts, M., & Heesterbeek, J. A. P. (2009).
+#' Early Epidemiological Assessment of the Virulence of Emerging Infectious
+#' Diseases: A Case Study of an Influenza Pandemic. PLOS ONE, 4(8), e6852.
+#' \doi{10.1371/journal.pone.0006852}
+#'
+#' @export
+#'
+#' @examples
+#' # get data pre-loaded with the package
+#' data("covid_data")
+#' df_covid_uk <- covid_data[covid_data$country == "United Kingdom", ]
+#'
+#' # estimate time varying severity without correcting for delays
+#' cfr_time_varying <- cfr_time_varying(
+#'   data = df_covid_uk,
+#'   burn_in = 7L
+#' )
+#' # View
+#' tail(cfr_time_varying)
+#'
+#' # estimate time varying severity while correcting for delays
+#' # obtain onset-to-death delay distribution parameters from Linton et al. 2020
+#' # view only the first values
+#' cfr_time_varying <- cfr_time_varying(
+#'   data = df_covid_uk,
+#'   delay_density = function(x) dlnorm(x, meanlog = 2.577, sdlog = 0.440),
+#'   burn_in = 7L
+#' )
+#' tail(cfr_time_varying)
+#'
+cfr_time_varying <- function(data,
+                             delay_density = NULL,
+                             burn_in = 7,
+                             smoothing_window = NULL) {
+  # input checking
+  # expect an integer-like number
+  checkmate::assert_int(burn_in, lower = 0)
+
+  # expect rows more than burn in value
+  checkmate::assert_data_frame(data, min.cols = 3, min.rows = burn_in + 1)
+  # check that input `<data.frame>` has columns date, cases, and deaths
+  checkmate::assert_names(
+    colnames(data),
+    must.include = c("date", "cases", "deaths")
+  )
+  # check for any NAs among data
+  checkmate::assert_data_frame(
+    data[, c("date", "cases", "deaths")],
+    any.missing = FALSE
+  )
+  checkmate::assert_number(smoothing_window, lower = 1, null.ok = TRUE)
+
+  stopifnot(
+    "`smoothing_window` must be an odd number greater than 0" =
+      (smoothing_window %% 2 != 0),
+    "`delay_density` must be a function with a single required argument,
+    and evaluating distribution density at a vector of values and returning a
+    numeric vector of the same length.
+    E.g. function(x) stats::dgamma(x = x, shape = 5, scale = 1)" =
+      (test_fn_req_args(delay_density) &&
+        test_fn_num_out(delay_density)) || is.null(delay_density)
+  )
+
+  # prepare a new dataframe with smoothed columns if requested
+  # all temporary operations are performed on df_temp,
+  # data is returned with only three new columns added, and no other changes
+  df_temp <- data
+  # smooth cases if requested
+  if (is.null(smoothing_window)) {
+    # set to 0 for internal use only --- see below
+    smoothing_window <- 0
```

Comment:
I don't understand how it would align better. I would have to check for zero values anyway as I don't want users to encounter a warning when passing `0` to `stats::runmed()`. Passing a default value of 1 (as it previously was) is also not great as the window size is used in other calculations.